### PR TITLE
UI - Add hop creation and peers mgmt

### PR DESCRIPTION
**Pull request by tanganellilore**
##### SUMMARY
As discussed with @TheRealHaoLiu and @fosterseth this PR is to add new features on UI for
- Modify add new instace UI witch checkbox for new param `peers_from_control_nodes` and multi popup choice for peers
- usage of `/api/v2/instances` patch for manage peers
- add associate/disassociate buttons on instance peers tab (visible only for hop and execution node)
- add edit page for instance with possibility of change description, port, enablement and peer to control node

Below some screenshots:

## New Add Instance UI
image

#### Choice peers
image

#### UI with peer selected
image

## New peers add/deassociate button
image

#### when you click on `Add ` (similar to add instance). If you click save peers will be added
image

#### If you want to remove, you can simply select and click on `Disassociate`. Popup warning will be showed
image

#### Details page have a `Edit` button
image

#### if you click on `Edit`, a similar view of adding page will be prompted. Served on `/instances/:id/edit`
image

##### ISSUE TYPE
 - New or Enhanced Feature

##### COMPONENT NAME
 - UI

##### AWX VERSION

```
N/A
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by fosterseth** [comment]
image
would be nice to show (read only) the "connect to control nodes" here. I do see it on the edit page

**Comment by tanganellilore** [comment]
added on last commit.

## Enabled:
image

## Disabled:
image

I also fixed one bug on UI for checkbox

**Comment by kialam** [comment]
This is great work - thank you for submitting this for review! Some initial notes:
1. Boolean fields are represented in the UI as label and label detail components. See example:
Screenshot 2023-05-01 at 10 17 05 AM
2. Fields that are not editable in a form should be greyed out - e.g. the node type on the Instance edit form should not be a selectable choice.
Screenshot 2023-05-01 at 10 24 55 AM
3. The `Add` button for peers list should be called `Associate` to keep in line with the `Disassociate` button. We have the same model for Instance Groups and Instances.
Screenshot 2023-05-01 at 10 25 11 AM
5. The Edit button should always be anchored on the left and the primary button.
Screenshot 2023-05-01 at 10 25 27 AM
6. Option fields should be grouped together like so:
Screenshot 2023-05-01 at 10 28 33 AM

**Comment by tanganellilore** [comment]
@kialam Thanks for reviews. I will answer per points:
1. Ok i will try to modify it (i used a chip in the detail view to be more "modern" but if this is not allowed il will remove it)
2. As discusse with @TheRealHaoLiu should be possibile change the type of node after creation, this is why was enable the modify. Let me know which modification it's required.
3. Agree and also I suppose to change add with associate.
4. I will do it
5. I will do the grouping

Tomorrow i will add one commit for 1,3,4,5.
For point 2 i will wait your feedback

**Comment by kialam** [comment]
@tanganellilore thank you. I spoke with @fosterseth and @TheRealHaoLiu and we want to disable changing node type in the edit form for now.

**Comment by tanganellilore** [comment]
Ok, no problem.
Tomorrow i will send new commit on this branch

**Comment by kialam** [comment]
Looking at the failed unit test: 
`InstancesLookup › should render successfully with instance_details for edit instance` - I don't see an Instances look up the edit form: 
```
{!isEdit && (
        <InstancesLookup
          helperTextInvalid={peersMeta.error}
          isValid={!peersMeta.touched || !peersMeta.error}
          ...
```

**Comment by tanganellilore** [comment]
Form is one for add and edit.
In edit part I removed the peers because are managed on peers tab.

Anyway, I will add it with using peers api of instance.

**Comment by tanganellilore** [comment]
@kialam new commit submitted with all fixing discussed. Please re-run test to check if something is missing

**Review by fosterseth — APPROVED** [review]

File: awx/ui/src/components/Lookup/InstancesLookup.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+import LookupErrorMessage from './shared/LookupErrorMessage';
+import FieldWithPrompt from '../FieldWithPrompt';
+
+const QS_CONFIG = getQSConfig('instances', {
+  page: 1,
+  page_size: 5,
+  order_by: 'hostname',
+});
+
+function InstancesLookup({
```
Comment (kialam) [inline awx/ui/src/components/Lookup/InstancesLookup.js]: I'm debating if this is better named PeersLookup. The merged query params and filtering make this a very specific subset of Instances being queried and is not merely a generic Instances Lookup.

File: awx/ui/src/components/StatusIcon/icons.js [hunk 2]
```diff
@@ -41,6 +41,7 @@ const icons = {
   skipped: MinusCircleIcon,
   canceled: ExclamationTriangleIcon,
   changed: ExclamationTriangleIcon,
+  enabled: CheckCircleIcon,
```
Comment (kialam) [inline awx/ui/src/components/StatusIcon/icons.js]: We can remove this

File: awx/ui/src/components/StatusLabel/StatusLabel.js [hunk 3]
```diff
@@ -24,6 +24,7 @@ const colors = {
   disabled: 'grey',
   canceled: 'orange',
   changed: 'orange',
+  enabled: 'green',
```
Comment (kialam) [inline awx/ui/src/components/StatusLabel/StatusLabel.js]: Same here

File: awx/ui/src/screens/Instances/InstanceDetail/InstanceDetail.js [hunk 4]
```diff
@@ -205,13 +206,41 @@ function InstanceDetail({ setBreadcrumb, isK8s }) {
             }
           />
           <Detail label={t`Node Type`} value={instance.node_type} />
+          <Detail label={t`Host`} value={instance.ip_address} />
+          <Detail label={t`Instance Port`} value={instance.listener_port} />
+          {(isExecutionNode || isHopNode) && (
+            <Detail
+              label={t`Connection to control nodes`}
+              value={
+                instance.peers_from_control_nodes ? t`Enabled` : t`Disabled`
```
Comment (kialam) [inline awx/ui/src/screens/Instances/InstanceDetail/InstanceDetail.js]:
```suggestion
                instance.peers_from_control_nodes ? t`On` : t`Off`
```

File: awx/ui/src/components/StatusLabel/StatusLabel.js (hunk 3 above)
Comment (kialam) [inline awx/ui/src/components/StatusLabel/StatusLabel.js]:
```suggestion
```

File: awx/ui/src/components/StatusLabel/StatusLabel.js [hunk 5]
```diff
@@ -53,6 +54,7 @@ export default function StatusLabel({ status, tooltipContent = '', children }) {
     disabled: t`Disabled`,
     canceled: t`Canceled`,
     changed: t`Changed`,
+    enabled: t`Enabled`,
```
Comment (kialam) [inline awx/ui/src/components/StatusLabel/StatusLabel.js]:
```suggestion
```

File: awx/ui/src/components/StatusLabel/StatusLabel.js [hunk 6]
```diff
@@ -104,6 +106,7 @@ StatusLabel.propTypes = {
     'disabled',
     'canceled',
     'changed',
+    'enabled',
```
Comment (kialam) [inline awx/ui/src/components/StatusLabel/StatusLabel.js]:
```suggestion
```

**Comment by tanganellilore** [comment]
@kialam removed unused lines and changed the label for peers.

Only open point for "instancelookup" but comment is above.

File: awx/ui/src/screens/Instances/InstanceEdit/InstanceEdit.test.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+const updatedInstance = {
+  node_type: 'hop',
+  peers: ['test-peer'],
+};
+
+describe('<InstanceEdit/>', () => {
+  let wrapper;
+  let history;
+
+  beforeAll(async () => {
```
Comment (kialam) [inline awx/ui/src/screens/Instances/InstanceEdit/InstanceEdit.test.js:81]:
```suggestion
  beforeEach(async () => {
```
^ fix the failing unit tests

**Comment by kialam** [comment]
The last unit test appears to have caught a bug:
```
<ErrorDetail__CardBody>
                                            <PFCardBody className="ErrorDetail__CardBody-sc-1grrpky-1 eUlxDj">
                                              <div className="pf-c-card__body ErrorDetail__CardBody-sc-1grrpky-1 eUlxDj">
                                                <strong>
                                                  TypeError
                                                  : 
                                                  Cannot read properties of undefined (reading &#39;data&#39;)
                                                </strong>
                                              </div>
                                            </PFCardBody>
                                          </ErrorDetail__CardBody>
                                          <ErrorDetail___StyledCardBody2>
                                            <PFCardBody className="ErrorDetail__CardBody-sc-1grrpky-1 ErrorDetail___StyledCardBody2-sc-1grrpky-2 eUlxDj cNBfAP">
                                              <div className="pf-c-card__body ErrorDetail__CardBody-sc-1grrpky-1 ErrorDetail___StyledCardBody2-sc-1grrpky-2 eUlxDj cNBfAP">
                                                TypeError: Cannot read properties of undefined (reading &#39;data&#39;)
                                                    at /Users/kilam/Sites/awx/awx/ui/src/screens/Instances/InstanceEdit/InstanceEdit.js:40:49
                                                    at processTicksAndRejections (node:internal/process/task_queues:96:5)
                                                    at /Users/kilam/Sites/awx/awx/ui/src/hooks/useRequest.js:32:28
                                              </div>
                                            </PFCardBody>
                                          </ErrorDetail___StyledCardBody2>
```

**Comment by kialam** [comment]
@tanganellilore trying to get your PR to the finish line. We have some unit test failures that need to be fixed, otherwise it's looking really good. The last bit is to just change `InstancesLookup` to `PeersLookup` and we should be ready to merge.

**Comment by kialam** [comment]
This should get the unit tests passing:
```diff
diff --git a/awx/ui/src/screens/Instances/InstanceEdit/InstanceEdit.test.js b/awx/ui/src/screens/Instances/InstanceEdit/InstanceEdit.test.js
index 64e2967f77..1447f35d61 100644
--- a/awx/ui/src/screens/Instances/InstanceEdit/InstanceEdit.test.js
+++ b/awx/ui/src/screens/Instances/InstanceEdit/InstanceEdit.test.js
@@ -74,10 +74,11 @@ describe('<InstanceEdit/>', () => {
   let wrapper;
   let history;
 
-  beforeAll(async () => {
+  beforeEach(async () => {
     useDebounce.mockImplementation((fn) => fn);
     history = createMemoryHistory();
     InstancesAPI.readDetail.mockResolvedValue({ data: instanceData });
+    InstancesAPI.readPeers.mockResolvedValue({ data: [] });
 
     await act(async () => {
       wrapper = mountWithContexts(
```

**Review by kialam — APPROVED** [review]
Excellent work. Once tests pass this is good to merge.

**Comment by fosterseth** [comment]
api-test, api-swagger, awx-collection, and api-scheme are expected to fail for now, as the feature branch will also fail those. I can address those later.

However the ui-lint and ui-test-screens tests should pass I think

**Comment by tanganellilore** [comment]
Test should be fixed (excluded @fosterseth tests).
Renamed file and disabled Instance type on edit page (instead disappear it)

**Pull request merged**